src/routes/docs/tutorials/pagination/step-7/+page.markdoc

```diff
@@ -0,0 +1,265 @@
+
+export function useTodos() {
+  return useContext(TodosContext);
+}
+
+export function TodosProvider(props) {
+  const [todos, setTodos] = useState([]);
+  const [offset, setOffset] = useState(0);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [counts, setCounts] = useState({
+    totalPages: 500,
+    totalResults: 1000,
+  });
+  const pageLimit = 4;
+  const hasNext = counts.totalPages > currentPage;
+
+  async function loadNext() {
+    if (hasNext) {
+      setOffset(offset + pageLimit);
+    }
+  }
+
+  async function loadPrev() {
+    if (offset >= pageLimit) {
+      setOffset(offset - pageLimit);
+    }
+  }
+
+  async function loadPage(pageNumber) {
+    setOffset((pageNumber - 1) * pageLimit);
+  }
+
+  async function init() {
+    try {
+      const response = await databases.listDocuments(
+        TODOS_DATABASE_ID,
+        TODOS_COLLECTION_ID,
+        [
+          Query.orderAsc("$createdAt"),
+          Query.limit(pageLimit),
+          Query.offset(offset),
+        ],
+      );
+
+      setTodos(response.documents);
+      setCurrentPage(Math.floor(offset / pageLimit) + 1);
+      setCounts({
+        totalPages: response.total / pageLimit,
+        totalResults: response.total,
+      });
+    } catch (error) {
+      console.log(error);
+    }
+  }
+  useEffect(() => {
+    init();
+  }, [offset]);
+
+  return (
+    <TodosContext.Provider
+      value={{
+        current: todos,
+        loadNext,
+        loadPrev,
+        loadPage,
+        currentPage,
+        totalPages: counts.totalPages,
+        hasNext,
+        hasPrev: currentPage > 1,
+      }}>
+      {props.children}
+    </TodosContext.Provider>
+  );
+}
+```
+In this updated context:
+ * We're using `Query.limit()` to set the number of tasks we want to fetch per page.
+ * We're using `Query.offset()` to set the starting point for fetching tasks based on the current page.
+ * We've added `loadNextPage`, `loadPrevPage`, and `loadPage` function to handle pagination.
+
+ Now, you can use this context in your components to display todo items and handle pagination using Appwrite's offset pagination methods.
+
+# Update the Todos page {% #update-the-todoss-page %}
+
+Now, let's use our Pagination component in the Todos page.
+{% tabs %}
+{% tabsitem #Basic-pagination title="Basic Pagination" %}
+Update the Todos page in `src/pages/Todos.jsx`:
+```js
+import BasicPagination from '../components/BasicPagination';
+import {useTodos} from "../lib/context/todos";
+export function TodosWithBasicPagination() {
+  const todos = useTodos();
+
+  return (
+    <section>
+      <h2>Todos</h2>
+      <ul>
+        {todos.current.map(todo => (
+          <li key={`todo-${todo.taskId}`}>
+            <span>{`Task-${todo.taskId}: `}</span>
+            <input type='checkbox' id={todo.taskId} name={todo.title} />
+            <label htmlFor={todo.title}>{todo.title}</label>
+          </li>
+        ))}
+      </ul>
+      <BasicPagination
```

Comment:
https://github.com/appwrite/website/assets/29069505/ca916780-1dc2-43b5-a498-12f6c0703d7e

Remember to debounce your buttons so I can't spam next and go past the last page!

Reply:
So rewriting the basic offset pagination component like this:
``` js
import React, { useCallback } from "react";

function BasicPagination(props) {
  const { loadPrev, loadNext, hasNext, hasPrev } = props;

  const debounce = (func, delay) => {
    let debounceTimer;
    return (...args) => {
      clearTimeout(debounceTimer);
      debounceTimer = setTimeout(() => func(...args), delay);
    };
  };

  const debouncedLoadPrev = useCallback(debounce(loadPrev, 300), [loadPrev]);

  const debouncedLoadNext = useCallback(debounce(loadNext, 300), [loadNext]);

  return (
    <div>
      <button disabled={!hasPrev} onClick={debouncedLoadPrev}>
        Previous
      </button>
      <button disabled={!hasNext} onClick={debouncedLoadNext}>
        Next
      </button>
    </div>
  );
}

export default BasicPagination;
```